File: src/main/java/com/uiptv/shared/Episode.java
package com.uiptv.shared;

import com.uiptv.model.Account;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.json.JSONObject;

import java.util.Map;

import static com.uiptv.util.StringUtils.getXtremeStreamUrl;
import static com.uiptv.util.StringUtils.safeGetString;

@Data
@NoArgsConstructor
public class Episode extends BaseJson {

    String id, episodeNum, title, containerExtension, custom_sid, added, season, direct_source, cmd;
    EpisodeInfo info;

    public Episode(Account account, Map map) {
        if (map == null) return;
        this.id = safeGetString(map, "id");
        this.episodeNum = safeGetString(map, "episode_num");
        this.title = safeGetString(map, "title");
        this.containerExtension = safeGetString(map, "container_extension");
        this.custom_sid = safeGetString(map, "custom_sid");
        this.added = safeGetString(map, "added");
        this.season = safeGetString(map, "season");
        this.direct_source = safeGetString(map, "direct_source");
        this.info = new EpisodeInfo((Map) map.get("info"));
        mergeEpisodeLevelArtwork(map);
        this.cmd = getXtremeStreamUrl(account, id, containerExtension);
    }

    public static Episode fromJson(String json) {
        try {
            JSONObject jsonObj = new JSONObject(json);
            Episode episode = new Episode();
            episode.setId(safeGetString(jsonObj, "id"));
            episode.setEpisodeNum(safeGetString(jsonObj, "episodeNum"));
            episode.setTitle(safeGetString(jsonObj, "title"));
            episode.setContainerExtension(safeGetString(jsonObj, "containerExtension"));
            episode.setCustom_sid(safeGetString(jsonObj, "custom_sid"));
            episode.setAdded(safeGetString(jsonObj, "added"));
            episode.setSeason(safeGetString(jsonObj, "season"));
            episode.setDirect_source(safeGetString(jsonObj, "direct_source"));
            episode.setCmd(safeGetString(jsonObj, "cmd"));
            if (jsonObj.has("info")) {
                episode.setInfo(new EpisodeInfo(jsonObj.getJSONObject("info").toMap()));
            }
            episode.mergeEpisodeLevelArtwork(jsonObj.toMap());
            return episode;
        } catch (Exception e) {
            return null;
        }
    }

    private void mergeEpisodeLevelArtwork(Map map) {
        if (map == null) return;
        if (this.info == null) {
            this.info = new EpisodeInfo();
        }
        String current = this.info.getMovieImage();
        if (!isBlankLike(current)) {
            return;
        }
        String rootEpisodeImage = firstNonBlank(
                map,
                "movie_image",
                "thumbnail",
                "still_path",
                "cover_big",
                "cover",
                "screenshot_uri",
                "stream_icon",
                "image",
                "poster"
        );
        if (!isBlankLike(rootEpisodeImage)) {
            this.info.setMovieImage(rootEpisodeImage);
        }
    }

    private String firstNonBlank(Map map, String... keys) {
        if (map == null || keys == null) return "";
        for (String key : keys) {
            String value = safeGetString(map, key);
            if (!isBlankLike(value)) return value.trim();
        }
        return "";
    }

    private boolean isBlankLike(String value) {
        if (value == null) return true;
        String v = value.trim();
        return v.isEmpty() || "null".equalsIgnoreCase(v) || "n/a".equalsIgnoreCase(v);
    }
}

File: server/src/main/kotlin/com/uiptv/shared/Episode.kt
package com.uiptv.shared

import com.uiptv.model.Account
import com.uiptv.util.StringUtils.getXtremeStreamUrl
import com.uiptv.util.StringUtils.safeGetString
import com.uiptv.util.json.KJsonObject

data class Episode @JvmOverloads constructor(
    var id: String? = null,
    var episodeNum: String? = null,
    var title: String? = null,
    var containerExtension: String? = null,
    @Suppress("PropertyName")
    var custom_sid: String? = null,
    var added: String? = null,
    var season: String? = null,
    @Suppress("PropertyName")
    var direct_source: String? = null,
    var cmd: String? = null,
    var info: EpisodeInfo? = null
) : BaseJson() {
    constructor(account: Account, map: Map<*, *>?) : this() {
        if (map == null) {
            return
        }
        id = safeGetString(map, "id")
        episodeNum = safeGetString(map, "episode_num")
        title = safeGetString(map, "title")
        containerExtension = safeGetString(map, "container_extension")
        custom_sid = safeGetString(map, "custom_sid")
        added = safeGetString(map, "added")
        season = safeGetString(map, "season")
        direct_source = safeGetString(map, "direct_source")
        info = EpisodeInfo(asStringMap(map["info"]))
        mergeEpisodeLevelArtwork(map)
        cmd = getXtremeStreamUrl(account, id, containerExtension)
    }

    companion object {
        @JvmStatic
        fun fromJson(json: String?): Episode? {
            if (json.isNullOrBlank()) {
                return null
            }
            return try {
                val jsonObj = KJsonObject(json)
                val episode = Episode(
                    id = safeGetString(jsonObj, "id"),
                    episodeNum = safeGetString(jsonObj, "episodeNum"),
                    title = safeGetString(jsonObj, "title"),
                    containerExtension = safeGetString(jsonObj, "containerExtension"),
                    custom_sid = safeGetString(jsonObj, "custom_sid"),
                    added = safeGetString(jsonObj, "added"),
                    season = safeGetString(jsonObj, "season"),
                    direct_source = safeGetString(jsonObj, "direct_source"),
                    cmd = safeGetString(jsonObj, "cmd")
                )
                val infoObject = jsonObj.optJSONObject("info")
                if (infoObject != null) {
                    episode.info = EpisodeInfo(infoObject.toMap())
                }
                episode.mergeEpisodeLevelArtwork(jsonObj.toMap())
                episode
            } catch (_: Exception) {
                null
            }
        }
    }

    private fun mergeEpisodeLevelArtwork(map: Map<*, *>?) {
        if (map == null) {
            return
        }
        if (info == null) {
            info = EpisodeInfo()
        }
        val current = info!!.movieImage
        if (!isBlankLike(current)) {
            return
        }
        val rootEpisodeImage = firstNonBlank(
            map,
            "movie_image",
            "thumbnail",
            "still_path",
            "cover_big",
            "cover",
            "screenshot_uri",
            "stream_icon",
            "image",
            "poster"
        )
        if (!isBlankLike(rootEpisodeImage)) {
            info!!.movieImage = rootEpisodeImage
        }
    }

    private fun firstNonBlank(map: Map<*, *>?, vararg keys: String): String {
        if (map == null) {
            return ""
        }
        for (key in keys) {
            val value = safeGetString(map, key)
            if (!isBlankLike(value)) {
                return value!!.trim()
            }
        }
        return ""
    }

    @Suppress("UNCHECKED_CAST")
    private fun asStringMap(value: Any?): Map<String, Any> {
        return if (value is Map<*, *>) {
            value as Map<String, Any>
        } else {
            emptyMap()
        }
    }

    private fun isBlankLike(value: String?): Boolean {
        if (value == null) {
            return true
        }
        val candidate = value.trim()
        return candidate.isEmpty() || candidate.equals("null", ignoreCase = true) || candidate.equals("n/a", ignoreCase = true)
    }
}
